### ISSUE TITLE
R inline output

### ISSUE THREAD

**Issue by jesusloplar** [nteract/hydrogen#1810]
I have been using python until now that I've got to use R. I like the way Hydrogen output Python results when executing (inline). So, I was wondering if the same could be done when using R (which outputs below).

**R output:**
<img width="108" alt="Screenshot 2019-11-01 at 22 30 05" src="https://user-images.githubusercontent.com/46565164/68057941-c7c1cf80-fcf7-11e9-85d3-b1f1adcfa91a.png">

**Python output:**
<img width="184" alt="Screenshot 2019-11-01 at 22 29 54" src="https://user-images.githubusercontent.com/46565164/68057959-d6a88200-fcf7-11e9-95a2-1f24cafc1ca7.png">

**Comment by wadethestealth** [comment]
The problem here is that R is probably also tacking on a newline at the end of execution text results. Instead of inline, can you run the R code above in the output area pane, and screenshot the result? This may help me confirm my theory, and if so, I would have to open a discussion about trimming new lines at the end of stream outputs.

**Comment by jesusloplar** [comment]
Presumably, this is what you asked for. Thanks you so much in advance.

<img width="726" alt="Screenshot 2019-11-01 at 23 50 55" src="https://user-images.githubusercontent.com/46565164/68060863-84b92980-fd02-11e9-9a95-17f156d904c4.png">

**Comment by wadethestealth** [comment]
Close. I am testing to see if running multiple of these concatenates in the output area, so I need you to run three different equations, but make sure you run them line by line not all at once. Thanks.

**Comment by jesusloplar** [comment]
Ah all right. Sorry for the misunderstanding. They are run line by line. Otherwise it would have been all of them in the same output box (not each one in a different output box, as shown in the first screen shot)

**Comment by wadethestealth** [comment]
That's unfortunate. Can you give me your kernel's name and version?

**Comment by wadethestealth** [comment]
The first screenshot is not what I am looking for. I am looking for that right pane like in your last screenshot, but I need you to run 3 equations not at the same time so I get at least 3 separate kernel results. Also can you let me know your kernel's name and version?

**Comment by jesusloplar** [comment]
This will help you to understand, hopefully.

<img width="440" alt="Screenshot 2019-11-02 at 17 28 08" src="https://user-images.githubusercontent.com/46565164/68074000-6866cc80-fd96-11e9-8edf-34268f6da4f3.png">

<img width="1306" alt="Screenshot 2019-11-02 at 17 10 45" src="https://user-images.githubusercontent.com/46565164/68073762-de1d6900-fd93-11e9-9fea-2e8954facd61.png">

**Comment by wadethestealth** [comment]
This isn't your kernel version. That's your R version. Which kernel do you use? Maybe this will help.
 https://stackoverflow.com/questions/50467766/in-irkernelinstallspec-jupyter-kernelspec-version-exits-with-code-one

But the other screenshot is exactly what I needed to know.

**Comment by jesusloplar** [comment]
Yepp, that one.

**Comment by wadethestealth** [comment]
Do you know which version you have though?

**Comment by jesusloplar** [comment]
I have been researching on how to prompt the version. As far as I have understood, since Catalina changed the shell to zsh, I haven't found any plausible command. Do you know?

**Comment by wadethestealth** [comment]
`jupyter kernelspec --version` should give me the necessary info.

**Comment by wadethestealth** [comment]
Even without it, I believe I have deduced the issue, but I want to be sure before I open an issue over at irkernel.

**Comment by jesusloplar** [comment]
I just realized that I didn't activate the virtual environment when running the command. My bad, sorry for the time.
`5.3.4`

**Comment by wadethestealth** [comment]
Alright. After a local installation and test here are the results sent back from an up-to-date IRKernel to Hydrogen after sending a request for `4+3`
```json
[
   "IO message:",
   {
      "idents" : [
         {
            "type" : "Buffer",
            "data" : [100,101,97,108,101,114,100,49,100,48,52,102,54,100,45,54,98,101,98,45,52,52,99,97,45,57,57,49,51,45,51,50,97,54,55,52,48,49,102,49,50,101]
         }
      ],
      "header" : {
         "msg_id" : "aee0e3da-fe6d-11e9-83a8-a9c397228e2e",
         "username" : "Wade",
         "session" : "00000000-0000-0000-0000-000000000000",
         "msg_type" : "display_data",
         "version" : "5.0"
      },
      "parent_header" : {
         "username" : "Wade",
         "session" : "00000000-0000-0000-0000-000000000000",
         "msg_type" : "execute_request",
         "msg_id" : "execute_f6808de1-e0c9-45d1-85df-edcab9815824",
         "date" : "2019-11-03T19:11:04.951Z",
         "version" : "5.0"
      },
      "metadata" : {},
      "content" : {
         "data" : {
            "text/html" : "7",
            "text/markdown" : "7",
            "text/latex" : "7",
            "text/plain" : "[1] 7"
         },
         "metadata" : {}
      },
      "buffers" : []
   }
]
```

You notice in the return, we get multiple text types (`text/html`, `text/markdown`, `text/latex`, `text/plain`). The current function for determining whether a result is a single line of text is as follows.
```javascript
export function isTextOutputOnly(data: Object) {
  const supported = React.Children.map(
    supportedMediaTypes.props.children,
    mediaComponent => mediaComponent.props.mediaType
  );
  const bundleMediaTypes = [...Object.keys(data)].filter(mediaType =>
    supported.includes(mediaType)
  );

  return bundleMediaTypes.length === 1 && bundleMediaTypes[0] === "text/plain"
    ? true
    : false;
}
```

Which essentially means, that if there is anything other than just a `text/plain` sent back its not a single line of text. It could be straight forward to just split the returned media types at the slash and make sure that all returned media types only begin with text regardless of what type of text this is. This would also inherently add single line support for markdown. Thoughts on improving the implementation in this manner @BenRussert @aviatesk?

**Comment by wadethestealth** [comment]
A second parsing idea has come to me. If `text/html` and `text/markdown` are equal I have guaranteed there are no markdown specifics in the text. Then if it is also equal to `text/latex` then I have guaranteed there is no html nodes in the text. This means I could only potentially have latex text if not plain text. Now, if I am special casing this for R only, then I can check if the text is now equal to the substring without the execution count of `text/plain`. This will guarantee I only have plain text and can be eligible for inline output.

**Comment by BenRussert** [comment]
@wadethestealth I'd be ok with a custom thing like you mention if you find something that works. A nice possible benefit from digging into this issue is that you may come up with a way to determine when html output (for IRKernel or any other kernel) should be rendered inline (e.g. if there are no html tags or newlines, or whatever you come up with.) I've seen the same basic thing happen in other kernels in some edge cases.

> Then if it is also equal to text/latex then I have guaranteed there is no html nodes in the text. 

We shouldn't need to worry about latex if there is html due to our [rendering priority](https://github.com/nteract/hydrogen/blob/master/lib/components/result-view/display.js#L41-L43). If you do decide you want to detect html tags, i'd try to find a proven solution on SO or a 3rd party library rather than rolling our own. Although, since security isn't the concern here, you could probably use a regex pattern you'd find on SO.

**Comment by wadethestealth** [comment]
@benrussert I could write a library for this, but currently I cannot find a 3rd party tool that doesn't render the HTML first. It's also hacky since it reads from current node properties which would require a ref

**Comment by wadethestealth** [comment]
Im just using the latex to verify there are no HTML nodes in `text/html` im not trying to render the latex.

**Comment by wadethestealth** [comment]
@BenRussert yes, but if there is a `<p>Hi.</p>` it is still one line with no styling. However we could just say it **has** to be raw text. But we shouldn't just check html, because priority may not always be the same in the future. There are lots of ways to break lines for each text we would have to catch. Because of all the specifics, I thought it would be better to special case and test if all texts are the exact same (minus the execution count) therefore I can just read it as normal to detect single line.
